### PR TITLE
LogFactory - Extracted config file loading and reloading logic

### PR DESCRIPTION
**Pull request by snakefoot**
Moved into dedicated config-loader

### CODE REVIEW

**Comment by ericnewton76** [comment]
I see you're actively working on it, and took basically the same track that I did.

**Comment by snakefoot** [comment]
> I see you're actively working on it, and took basically the same track that I did.

Yes but this PR limits itself to non-breaking changes. Allowing it to be included in NLog 4.6

> Any method that calls down to System.IO.File is basically a problem. That should exist basically in one place... 

One could also create a mock-interface for File-access that can be given to all classes where file-access makes sense. By default they will use the singleton implementation that performs the actual file-access, unless one is provided from the outside. And also making it possible to replace the singleton-instance dependending on platform.

**Comment by 304NotModified** [comment]
related: #3056

**Comment by snakefoot** [comment]
> @304NotModified related: #3056

Think @ericnewton76 was planning to make a major breaking change for NLog 5.0. This PR is "just" a refactoring without breaking changes.

**Comment by 304NotModified** [comment]
> XmlLoggingConfiguration actually shouldnt have any LogFactory reference at all... creates a circular reference where potentially its never vacated due to event hooks/etc. Usually nota problem ironically for but is probable memory leak.

Agree on that. Circular reference are bad.

> Also it sounds like the refactoring you are working on is huge. I prefer making refactoring in small steps with frequent commits.

I also really prefer that, to review things in a short term ;)

File: src/NLog/LogFactory.cs [hunk 1]
```diff
@@ -1077,7 +862,12 @@ public IEnumerable<string> GetCandidateConfigFilePaths()
                 return _candidateConfigFilePaths.AsReadOnly();
             }
 
-            return GetDefaultCandidateConfigFilePaths();
+            if (_configLoader is LoggingConfigurationFileLoader configFileLoader)
```
[the remark this answers is not in the document]
Comment (snakefoot) [reply src/NLog/LogFactory.cs]: LogFactory should not worry about candidate paths, and code should be removed. Keeping the code-smell as a reminder.

File: src/NLog/LogFactory.cs [hunk 2]
```diff
@@ -141,16 +125,22 @@ public LogFactory(LoggingConfiguration config)
...
-        /// <param name="config"></param>
-        /// <param name="file"></param>
-        internal LogFactory(LoggingConfiguration config, IFile file) : this(config)
+        /// <param name="configLoader">The config loader</param>
+        internal LogFactory(ILoggingConfigurationLoader configLoader)
         {
-            _file = file;
+            _configLoader = configLoader;
+#if !SILVERLIGHT && !__IOS__ && !__ANDROID__ && !NETSTANDARD1_3
+            if (configLoader is LoggingConfigurationFileReloader configFileReloader)
```
[the remark this answers is not in the document]
Comment (snakefoot) [reply src/NLog/LogFactory.cs]: Refactored the interface, and add a new method `Activated`

File: src/NLog/LogFactory.cs [hunk 3]
```diff
@@ -1285,121 +994,13 @@ private static Logger CreateDefaultLogger(ref LoggerCacheKey cacheKey)
...
-#else
-                if (File.Exists(configFile))
-                {
-                    config = LoadXmlLoggingConfiguration(configFile);
-                    return true;    // File exists, and maybe the config is valid, stop search
-                }
-#endif
-            }
-            catch (IOException ex)
+            if (_configLoader is LoggingConfigurationFileLoader configFileLoader)
```
[the remark this answers is not in the document]
Comment (snakefoot) [reply src/NLog/LogFactory.cs]: LogFactory should not perform explicit loading of config-files, and code should be removed. Keeping the code-smell as a reminder.

File: src/NLog/Config/LoggingConfigurationFileReloader.cs [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+
+namespace NLog.Config
+{
+    using System;
+    using System.Linq;
+    using System.Threading;
+    using NLog.Common;
+    using NLog.Internal;
+
+    internal class LoggingConfigurationFileReloader : LoggingConfigurationFileLoader
```
[the remark this answers is not in the document]
Comment (snakefoot) [reply src/NLog/Config/LoggingConfigurationFileReloader.cs]: Added comment with reason

File: src/NLog/LogFactory.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (304NotModified) [reply src/NLog/LogFactory.cs]: it's still an check on the class `LoggingConfigurationFileReloader `? (and not the interface)

File: src/NLog/LogFactory.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (304NotModified) [reply src/NLog/LogFactory.cs]: please add some todo comments for reminders. Code smells aren't easy to find.

File: src/NLog/LogFactory.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (304NotModified) [reply src/NLog/LogFactory.cs]: same as above, please add a TODO comment

**Comment by snakefoot** [comment]
@304NotModified Renamed classes and added TODO comment.

**Comment by snakefoot** [comment]
@304NotModified Guess build should be restarted after merge of #3077

File: src/NLog/Config/LoggingConfigurationLoaderFileWatcher.cs [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+    using System.Linq;
+    using System.Threading;
+    using NLog.Common;
+    using NLog.Internal;
+
+    /// <summary>
+    /// Enables FileWatcher for the currently loaded NLog Configuration File,
+    /// and supports automatic reload on file modification.
+    /// </summary>
+    internal class LoggingConfigurationLoaderFileWatcher : LoggingConfigurationLoaderFile
```
Comment (304NotModified) [inline src/NLog/Config/LoggingConfigurationLoaderFileWatcher.cs]: I'm sorry, but I think these class names are confusing.

It's from the names not clear those two classes are ConfigurationLoaders, and that this class is a specilized version of the "normal " file loader.

 I know those are internal, but will also popup in stacktraces and we need to maintain them ;)

proposal, but other proposals also accepted:

- LoggingConfigurationFileLoader
- LoggingConfigurationWatchableFileLoader

Maybe another option is to create "sub" namespaces to help in the naming.

Comment (snakefoot) [reply src/NLog/Config/LoggingConfigurationLoaderFileWatcher.cs]: Made one more rename attempt:

- LoggingConfigurationFileLoader
- LoggingConfigurationFileWatchingLoader

Comment (snakefoot) [reply src/NLog/Config/LoggingConfigurationLoaderFileWatcher.cs]: Renamed classes as proposed.

**Pull request merged**